### PR TITLE
HPCC-17063 Reduce compile time (in ScopedDependentTransformer) for some queries

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately

## Testing:

Ran full regression suite - no differences in generated code.  Saves approx 6 minutes processing time (6m out of 400m).  mvr.eclxml reduced from 5mins to 1m15 and memory from 14Gb to 2.3Gb.

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review.

File: ecl/hql/hqltrans.cpp [hunk 1]
```diff
@@ -4564,28 +4564,50 @@ void ScopedDependentTransformer::pushChildContext(IHqlExpression * expr, IHqlExp
     //NB: For this transformer it is the tables that are in scope that matter
     IHqlExpression * scope = expr->queryNormalizedSelector(false);
     //NB: Do no call createComma because that calls createDataset which unwinds a comma list!
```
Comment (shamser) [inline ecl/hql/hqltrans.cpp]: @ghalliday May be worth moving this comment before line 4586

File: ecl/hql/hqltrans.cpp [hunk 2]
```diff
@@ -4564,28 +4564,50 @@ void ScopedDependentTransformer::pushChildContext(IHqlExpression * expr, IHqlExp
...
-        childScope.setown(createValue(no_comma,LINK(scope), childScope.getClear()));
-    else
-        childScope.set(scope);
+    {
+        IHqlExpression * cur = childScope;
+        while (cur->getOperator() == no_comma)
+        {
+            IHqlExpression * prev = cur->queryChild(0);
+            if (prev == scope)
+                identical = true;
```
Comment (shamser) [inline ecl/hql/hqltrans.cpp]: @ghalliday Not sure why the loop is continuing once identical is set to 'true'..could it end the loop once identical==true?

**Review by shamser — COMMENTED** [review]
@ghalliday Code reviewed.. please see comments.

File: ecl/hql/hqltrans.cpp (hunk 2 above)
Comment (ghalliday) [inline ecl/hql/hqltrans.cpp]: yes, it could break.  The list is typically very short, but I will add it.

**Comment by ghalliday** [comment]
@richardkchapman updated following Shamser's comments

**Pull request merged**